resources/shaders/dxr_functions.hlsl

```diff
@@ -33,4 +33,34 @@ float3 unpack_position(float2 uv, float depth, float4x4 inv_vp)
 	return (wpos.xyz / wpos.w).xyz;
 }
 
-#endif //__DXR_FUNCTIONS_HLSL__
\ No newline at end of file
+float3 CalcPeturbedNormal(float3 normal, float3 normal_map, float3 tangent, float3 bitangent, float3 V, out float3 world_normal)
+{
+	float3x4 object_to_world = ObjectToWorld3x4();
+	float3 N = normalize(mul(object_to_world, float4(normal, 0)));
+	float3 T = normalize(mul(object_to_world, float4(tangent, 0)));
+#define CALC_B
+#ifndef CALC_B
+	const float3 B = normalize(mul(object_to_world, float4(bitangent, 0)));
+#else
+	T = normalize(T - dot(T, N) * N);
+	float3 B = cross(N, T);
+#endif
+	const float3x3 TBN = float3x3(T, B, N);
+
+	//float3 fN = N;
+	float3 fN = normalize(mul(normal_map, TBN));
+	fN = lerp(fN, -fN, dot(fN, V) < 0);
+
+	world_normal = N;
+
+	return fN;
+}
+
+float3 CalcPeturbedNormal(float3 normal, float3 normal_map, float3 tangent, float3 bitangent, float3 V)
```

Comment:
Why use temp here? Isn't it cleaner to just pass "0" immediately?

Reply:
can't do that since it is an out value